src/main/scala/dspjunctions/JtagAxi.scala

```diff
@@ -62,65 +68,60 @@ class JtagAxiMaster()(implicit p: Parameters) extends NastiModule()(p) {
     val ctrlAxi = new NastiIO
     val dataAxi = new NastiIO
   })
-
-  val tap = Module(new JtagTwoAxiMasterClocked(io.jtag.TCK.asClock, false.B)) // io.jtag.TRST))
-  io.jtag <> tap.io.jtag
-  tap.io.control.fsmAsyncReset := false.B
-
-  io.ctrlAxi <> tap.io.ctrlAxi
-  io.dataAxi <> tap.io.dataAxi
-}
-
-class JtagTwoAxiMasterClocked(modClock: Clock, modReset: Bool)(implicit val p: Parameters)
-  extends Module(override_clock=Some(modClock), override_reset=Some(modReset))
-  with HasNastiParameters {
-  
   val irLength = 4
+  val idcode = (BigInt(14), JtagIdcode(1, 1, 0x42))
+
+  val jtagClock = io.jtag.TCK.asClock
+  val jtagReset = Wire(Bool()) // reset=io.jtag.TRST
+
+  val (innerCtrlAxi, innerDataAxi) = withClockAndReset(clock = jtagClock, reset=jtagReset) {
+    val innerCtrlAxi = Wire(new NastiIO)
+    val innerDataAxi = Wire(new NastiIO)
+    val ctrl_aw_chain = Module(new DecoupledSourceChain(new NastiWriteAddressChannel))
+    val ctrl_w_chain  = Module(new DecoupledSourceChain(new NastiWriteDataChannel))
+    val ctrl_b_chain  = Module(new DecoupledSinkChain(new NastiWriteResponseChannel))
+    val ctrl_ar_chain = Module(new DecoupledSourceChain(new NastiReadAddressChannel))
+    val ctrl_r_chain  = Module(new DecoupledSinkChain(new NastiReadDataChannel))
+
+    val data_aw_chain = Module(new DecoupledSourceChain(new NastiWriteAddressChannel))
+    val data_w_chain  = Module(new DecoupledSourceChain(new NastiWriteDataChannel))
+    val data_b_chain  = Module(new DecoupledSinkChain(new NastiWriteResponseChannel))
+    val data_ar_chain = Module(new DecoupledSourceChain(new NastiReadAddressChannel))
+    val data_r_chain  = Module(new DecoupledSinkChain(new NastiReadDataChannel))
+
+    val tapIO = JtagTapGenerator(irLength, Map(
+      0 -> ctrl_aw_chain,
+      1 -> ctrl_w_chain,
+      2 -> ctrl_b_chain,
+      3 -> ctrl_ar_chain,
+      4 -> ctrl_r_chain,
+      5 -> data_aw_chain,
+      6 -> data_w_chain,
+      7 -> data_b_chain,
+      8 -> data_ar_chain,
+      9 -> data_r_chain
+    ), idcode = Some(idcode))
+
+    io.jtag <> tapIO.jtag
+    tapIO.control.fsmAsyncReset := false.B
+    jtagReset := tapIO.output.reset
```

Comment:
Yeah.

The root of the issue is that the JTAG FSM has a test-logic-reset state, which should reset all test logic. Since our chains are outside the JTAG TAP, but the TAP is in the same clock/reset block, this funky threading needs to happen...